src/render/gif.rs

```diff
@@ -0,0 +1,277 @@
+                }
+            }
+            Err(_) => {
+                log::error!("The gif file can't be used, using the default placeholder.");
+
+                Self::new_error(origin, stop)
+            }
+        }
+    }
+
+    pub fn new_error(origin: Point, stop: Point) -> Self {
+        Self::new_u8(origin, stop, GIF_MISSING)
+    }
+
+    pub fn new_u8(origin: Point, stop: Point, u8_array: &[u8]) -> Self {
+        let gif_height = stop.y - origin.y;
+        let gif_width = stop.x - origin.x;
+
+        let mut decoder = gif::DecodeOptions::new();
+
+        decoder.set_color_output(gif::ColorOutput::RGBA); //TODO we're repeating a those lines, maybe make a function (don't ask me how)
+
+        let mut decoder = decoder.read_info(u8_array).unwrap();
+
+        let mut decoded_frames = Vec::new();
+        let mut delays = Vec::new();
+
+        // Read all the frames in the u8 array.
+        while let Some(frame) = decoder.read_next_frame().unwrap() {
+            decoded_frames.push(Self::read_frame(frame, gif_height, gif_width));
+            delays.push(frame.delay);
+        }
+        Self {
+            stop,
+            origin,
+            decoded_frames,
+            current_frame: AtomicUsize::new(0),
+            delays,
+            time_frame_last_update: Rc::new(RefCell::new(Instant::now())),
+        }
+    }
+
+    pub fn draw(&self, target: &mut FrameBuffer) -> bool {
+        let frame = self.current_frame.load(Ordering::Relaxed);
+
+        //get the data for the specified frame
+        let frame_data = &self.decoded_frames[frame];
+
+        //convert the data to an ImageRaw
+        let raw_gif_frame =
+            ImageRaw::<BinaryColor>::new(&frame_data, (self.stop.x - self.origin.x) as u32);
+
+        //draw the ImageRaw on the buffer
+        let _ = Image::new(&raw_gif_frame, self.origin).draw(target);
+
+        //detect if we should change the frame
+        let last_display_time = self.time_frame_last_update.borrow().clone();
+        let current_time = Instant::now();
+        let elapsed_time = current_time - last_display_time;
+
+        if elapsed_time >= Duration::from_millis(u64::from(self.delays[frame] * 10)) {
+            //the delays in gifs are in increments of 10 ms
+            //https://docs.rs/gif/latest/gif/struct.Frame.html#structfield.delay
+
+            //update the variable only if we update the frame
+            *self.time_frame_last_update.borrow_mut() = current_time;
+
+            //increment the current_frame using atomic operations
+            let next_frame = frame + 1;
+
+            let has_gif_ended = next_frame >= self.decoded_frames.len();
+            if has_gif_ended {
+                //reset to frame 0
+                self.current_frame.store(0, Ordering::Relaxed);
+            } else {
+                self.current_frame.store(next_frame, Ordering::Relaxed);
+            }
+            return has_gif_ended;
+        }
+        return false;
```

Comment:
this can just be `false`, the return is unnecessary

Reply:
oh yes, it's just that I normally work in javascript and java, so returns have become an automatism for me, will fix shortly